### DISCUSSION TITLE
offhours instance filtering

### DISCUSSION THREAD

**Discussion by Mykykyta** [cloud-custodian/cloud-custodian#7461] [AWS] [answered]
Tried literally everything and can't make my onhours\ofhours policy see the instance. What I do wrong?

Here are the ec2 instance tags I have on my aws

![image](https://user-images.githubusercontent.com/32738499/171492305-beb5876b-dd12-4c98-9eb1-93f88829ecbc.png)

My policy

![image](https://user-images.githubusercontent.com/32738499/171492106-703fe2c4-d0d4-4bb2-94dd-ceead879ac4d.png)

I run the policy locally and can stop\start the instance manually using cloud custodian actions.

`custodian run -c policy-offhours-ec2.yml -s .`

And it logs

`2022-06-01 21:41:50,853: custodian.filters:WARNING offhour implicitly filtered 23 of 27 resources key:State.Name on running`
` 2022-06-01 21:41:50,853: custodian.policy:INFO policy:ec2-offhours resource:ec2 region:us-east-1 count:0 time:2.08`

**Comment by thisisshi** [comment, marked as answer]
Hi @Mykykyta there's a few things you need to consider when writing an off hours policy:
1. You can set the default off hour in your policy by setting the offhour attribute in the filter, what this means is that when the policy is run at that hour any resources that are applicable will be filtered in.
2. You can set the schedule at the resource level using the `maid_offhour` tag, or a certain tag that you specify with the `tag` attribute. In this tag, you can specify when the instance should be turned off but you will also need to run the policy during that hour to actually turn the instance off. For example, if your schedule looks like so: `off=(M-F,18)` this means that the instance will be turned off Monday - Friday at 6 PM. In order to actually turn the instance off, you need to make sure you are running the policy at 6 PM each day. The same logic applies to the onhour filter as well.
3. As mentioned in (2), you need to make sure you're running this policy at the hour that's specified on the resource tag. In reality, you should run this policy each hour to ensure that all resources that have a schedule applied in their tags are being respected, you can do this by running the policy as a cron job, for example.

**Reply by Mykykyta** [reply]
Thank you, it helped!

**Reply by Mykykyta** [reply]
Although I am unable to start the instance with `maid_offhour: on=(M-F,21)` or `maid_offhour: off=(M-F,19);on=(M-F,21)` while running it at 21 42. Can I start\stop the instance only once per 1 hour using cloud custodian? Are there any restrictions like that?

![image](https://user-images.githubusercontent.com/32738499/171768963-c11b84db-a989-4315-b33e-900f0760fb47.png)
![image](https://user-images.githubusercontent.com/32738499/171768900-56dcd405-e112-4cce-b3a1-9c5547b9d95c.png)
![image](https://user-images.githubusercontent.com/32738499/171770782-13f2e617-1727-474e-a040-63ba06ff2fec.png)

**Reply by thisisshi** [reply]
Assuming the log is showing UTC, 03:42 UTC is actually 11:42 PM Eastern or 23:42 Eastern, not 21:42. The filter will also implicitly filter out resources that are not stopped but I think the issue is the time that you are running the policy.

**Reply by Mykykyta** [reply]
Unfortunately, the log was showing GMT+2. I have started and stopped it again successfully at different hours. For some reason, I can't start and stop the instance during the same hour. That was my problem. Thank you so much for your help!
![Screencast-from-03 06 2022-17_10_00](https://user-images.githubusercontent.com/32738499/171953950-635271f3-0977-4dec-912a-f9e648d84355.gif)
